reporter/src/main/kotlin/reporters/CycloneDxReporter.kt

```diff
@@ -140,61 +115,116 @@ class CycloneDxReporter : Reporter {
-                "URL to the ${ortResult.repository.vcsProcessed.type} repository of the projects"
+                vcs.url,
+                "URL to the ${vcs.type} repository of the projects"
             )
-        }
 
-        ortResult.getPackages().forEach { (pkg, _) ->
-            // TODO: We should actually use the concluded license expression here, but we first need a workflow to
-            //       ensure it is being set.
-            val declaredLicenseNames = pkg.declaredLicensesProcessed.allLicenses
-            val detectedLicenseNames = ortResult.getDetectedLicensesForId(pkg.id, input.packageConfigurationProvider)
+            input.ortResult.getPackages().forEach { (pkg, _) ->
+                addPackageToBom(input, pkg, bom)
+            }
 
-            val licenseObjects = mapLicenseNamesToObjects(declaredLicenseNames, "declared license", input) +
-                    mapLicenseNamesToObjects(detectedLicenseNames, "detected license", input)
+            writeBomToFile(bom, outputFile)
+            outputFiles += outputFile
+        } else {
+            projects.forEach { project ->
+                val reportFilename = "$REPORT_BASE_FILENAME-${project.id.toPath("-")}.$REPORT_EXTENSION"
+                val outputFile = outputDir.resolve(reportFilename)
 
-            val binaryHash = mapHash(pkg.binaryArtifact.hash)
-            val sourceHash = mapHash(pkg.sourceArtifact.hash)
+                val bom = Bom().apply { serialNumber = "urn:uuid:${UUID.randomUUID()}" }
 
-            val (hash, purlQualifier) = if (binaryHash == null && sourceHash != null) {
-                Pair(sourceHash, "?classifier=sources")
-            } else {
-                Pair(binaryHash, "")
+                // Add information about projects as external references at the BOM level.
+                bom.addExternalReference(
+                    ExternalReference.Type.VCS,
+                    project.vcsProcessed.url,
+                    "URL to the project's ${project.vcsProcessed.type} repository"
+                )
+
+                bom.addExternalReference(ExternalReference.Type.WEBSITE, project.homepageUrl)
+
+                val licenseNames = project.declaredLicensesProcessed.allLicenses +
+                        input.ortResult.getDetectedLicensesForId(project.id, input.packageConfigurationProvider)
+
+                bom.addExternalReference(ExternalReference.Type.LICENSE, licenseNames.joinToString(", "))
+
+                bom.addExternalReference(ExternalReference.Type.BUILD_SYSTEM, project.id.type)
+                bom.addExternalReference(
+                    ExternalReference.Type.OTHER,
+                    project.id.toPurl(),
+                    "Package-URL of the project"
+                )
+
+                val dependencies = project.collectDependencies()
+                val packages = input.ortResult.getPackages().mapNotNull { (pkg, _) ->
+                    pkg.takeIf { it.id in dependencies }
```

Comment:
Consider using filter instead.

Reply:
With `filter` I'd still need a separate `map` before or after, iterating over the packages two times in total. This way I only need to iterate once.